api/src/main/java/org/openmrs/api/impl/ConceptServiceImpl.java

```diff
@@ -332,6 +332,61 @@ public void purgeDrug(Drug drug) throws APIException {
 	public Concept getConcept(Integer conceptId) throws APIException {
 		return dao.getConcept(conceptId);
 	}
+
+	/**
+	 * @see org.openmrs.api.ConceptService#getConceptByReference(String conceptRef)
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Concept getConceptByReference(String conceptRef) {
+		if (StringUtils.isBlank(conceptRef)) {
+			return null;
+		}
+		
+		Concept cpt;
+		//check if input is a valid Uuid
+		if (isValidUuidFormat(conceptRef)) {
+			cpt = Context.getConceptService().getConceptByUuid(conceptRef);
+			if (cpt != null) {
+				return cpt;
+			}
+		}
+
+		//handle mapping
+		int idx = conceptRef.indexOf(":");
+		if (idx >= 0 && idx < conceptRef.length() - 1) {
+			String conceptSource = conceptRef.substring(0, idx);
+			String conceptCode = conceptRef.substring(idx + 1);
+			cpt = Context.getConceptService().getConceptByMapping(conceptCode, conceptSource);
+			if (cpt != null) {
+				return cpt;
+			}
+		}
+		
+		//handle id
+		int conceptId = NumberUtils.toInt(conceptRef, -1);
+		if (conceptId >= 0) {
+			cpt = Context.getConceptService().getConcept(conceptId);
+		} else {
+			//handle name
+			cpt = Context.getConceptService().getConceptByName(conceptRef);
+		}
+		
+		if (cpt != null) {
+			return cpt;
+		}
+
+		if (conceptRef.contains(".")) {
+			try {
+				return Context.getConceptService().getConceptByReference(evaluateStaticConstant(conceptRef));
+			}
+			catch (IllegalArgumentException e) {
+				log.error("Error while evaluating '{}' as a constant", conceptRef, e);
```

Comment:
How will the caller know that something went wrong?

Reply:
@dkayiwa I have been updating all my pending issues, but honestly speaking, I have failed to understand what is required here.
cc @tendomart @ibacher

Reply:
I think this is mostly fine, but I'd move the logging level to `warn` rather than `error`. This method is kind of speculative (try to find a way to figure out what concept this is) so it's not *really* an error if it's not a static reference. Also, the message logged here seems like it would be a better message to attach to the exception.

@dkayiwa I don't know that there's value in throwing the exception from here. In most cases, this method would probably be called to see if a string can (somehow) be resolved to a concept and if it can't, `null` seems like an appropriate value. An error most likely only indicates that a string like "MAP:1234" is not a Java constant as well as not matching any of the previous exit conditions.

Reply:
Sounds fair enough! 😊

Reply:
I have made more changes and switched to `log.warn` and many more

Thanks @ibacher for the detailed explanation..